include/shark/Algorithms/DirectSearch/Indicators/HypervolumeIndicator.h

```diff
@@ -29,128 +29,70 @@
-	* \param [in] referencePoint reference for measuring the hypervolume
-	*/
-	template<typename Extractor, typename ParetoFrontType, typename VectorType>
-	std::size_t leastContributor( Extractor const& extractor, ParetoFrontType const& front, VectorType const& referencePoint)
+	/// \brief Determines the point contributing the least hypervolume to the overall front of points.
+	///
+	/// This version uses the reference point estimated by the last call to updateInternals.
+	///
+	/// \param [in] front pareto front of points
+	template<typename ParetoFrontType>
+	std::size_t leastContributor( ParetoFrontType const& front)
 	{
 		std::vector<double> indicatorValues( front.size() );
 		SHARK_PARALLEL_FOR( int i = 0; i < static_cast< int >( front.size() ); i++ ) {
-
-			HypervolumeIndicator ind(*this);
-
-			ParetoFrontType copy( front );
+			std::vector<RealVector> copy( front.begin(), front.end() );
 			copy.erase( copy.begin() + i );
-
-			indicatorValues[i] = ind( extractor, copy,referencePoint);
+			
+			HypervolumeCalculator hv;
+			indicatorValues[i] = hv(copy,m_reference);
 		}
 
 		std::vector<double>::iterator it = std::max_element( indicatorValues.begin(), indicatorValues.end() );
 
-		return std::distance( indicatorValues.begin(), it );
-	}
-	
-	/**
-	 * \brief Determines the point contributing the least hypervolume to the overall front of points.
-	 *
-	 * This version uses the reference point estimated by the last call to updateInternals.
-	 * 
-	 * \param [in] extractor Extracts point information from front elements.
-	 * \param [in] front pareto front of points
-	 */
-	template<typename Extractor, typename ParetoFrontType>
-	std::size_t leastContributor( Extractor const& extractor, ParetoFrontType const& front)
-	{
-		return leastContributor(extractor,front,m_reference);
+		return it - indicatorValues.begin();
 	}
 	
 	/// \brief Updates the internal variables of the indicator using a whole population.
 	///
 	/// Calculates the reference point of the volume from the population
 	/// using the maximum value in every dimension+1
-	/// \param extractor Extracts point information from set.
-	/// \param set The set of points.
-	template<typename Extractor, typename PointSet>
-	void updateInternals(Extractor const& extractor, PointSet const& set){
+	///
+	/// \param [in] front pareto front of points
+	template<typename ParetoFrontType>
+	void updateInternals( ParetoFrontType const& front){
 		m_reference.clear();
-		if(set.empty()) return;
+		if(front.empty()) return;
 		
 		//calculate reference point
-		std::size_t noObjectives = extractor(set[0]).size();
+		std::size_t noObjectives = front[0].size();
 		m_reference.resize(noObjectives);
-		m_reference.clear();
 		
-		for( unsigned int i = 0; i < set.size(); i++ )
-			noalias(m_reference) = max(m_reference, extractor(set[i]));
+		for(auto const& point: front)
+			noalias(m_reference) = max(m_reference, point);
 		
 		noalias(m_reference) += 1.0;
```

Comment:
I still think that 1 is too much of a magic value. There needs to be some value, but the user should be responsible for setting it explicitly. This is like the C = 1 default for an SVM. Really, some parameters need a design choice and cannot have a default. If at all, then the default should be relative to the range of observed values, e.g., param * (max - min), with say, param=1.